### PR TITLE
[MOG-10]: Implement Download Streaming

### PR DESCRIPTION
**Pull request by Maiz27**
<h2>Greptile Overview</h2>

<h3>Greptile Summary</h3>

This PR implements server-side streaming download for image collections, replacing the previous client-side Web Worker approach. The implementation adds a new `/api/download/stream` endpoint that generates ZIP files on-demand with intelligent caching based on content hashing.

**Key Changes:**
- New streaming API endpoint with two-phase flow (prepare → download) for better UX during generation
- Content-based cache invalidation using MD5 hashing of gallery items to ensure fresh zips when Sanity content changes
- Disk caching with 1-hour TTL stored in OS temp directory with automatic cleanup
- Backpressure control (1MB highWaterMark) and batch processing (3 images per batch) to prevent memory exhaustion
- Windows file lock retry logic with exponential backoff for rename operations
- SSRF protection validating all URLs are from `cdn.sanity.io` hostname
- Comprehensive test suite using Vitest covering cache invalidation, retry logic, and cleanup
- Client hook with content-type validation to catch HTML error pages before blob parsing
- New `DownloadDrawer` component with multi-step UX flow

**Architecture Improvements:**
- Uses `mkdtemp` for atomic file generation preventing race conditions
- Stream-based image processing avoiding loading full images into RAM
- Proper abort signal propagation for client disconnects
- Lock file debouncing (5 min) to prevent concurrent cleanup scans

<h3>Confidence Score: 4/5</h3>

- This PR is safe to merge with minor monitoring recommended for production performance
- The implementation is well-architected with comprehensive error handling, security validations (SSRF protection, URL whitelist), proper resource management (backpressure control, batch processing), and good test coverage. Previous review threads show extensive iteration addressing race conditions, memory issues, and edge cases. The code demonstrates production-ready patterns including content-based cache invalidation, atomic file operations via mkdtemp, exponential backoff for Windows locks, and proper stream cleanup. Score is 4/5 rather than 5/5 due to the complexity of the streaming implementation and the need to monitor cleanup performance and cache hit rates in production.
- No files require special attention - the implementation follows best practices throughout

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/app/api/download/stream/route.ts | New streaming download endpoint with caching, content-based invalidation, and comprehensive error handling. Implements backpressure control and Windows lock retry logic. |
| src/app/api/download/stream/route.test.ts | Comprehensive test coverage for streaming route including cache invalidation, Windows lock retry, and cleanup logic validation. |
| src/lib/hooks/useDownloadCollection.ts | Added `downloadStream` method with two-phase prepare/download flow, content-type validation, and improved error handling. |
| src/components/drawers/DownloadDrawer.tsx | New drawer component with multi-step flow (email → choice → stream/parts), clean UX for stream preparation state. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant User
    participant DownloadDrawer
    participant useDownloadCollection
    participant API as /api/download/stream
    participant Sanity
    participant FileSystem
    participant Archive as Archiver

    User->>DownloadDrawer: Enter email
    DownloadDrawer->>DownloadDrawer: Validate email
    User->>DownloadDrawer: Click "Download Full Collection"
    DownloadDrawer->>useDownloadCollection: downloadStream(email)
    
    Note over useDownloadCollection,API: PREPARE Phase
    useDownloadCollection->>API: POST (mode: prepare)
    API->>API: Validate request & auth (private collections)
    API->>Sanity: Fetch collection metadata & items
    Sanity-->>API: Return gallery items & title
    API->>API: Generate content hash (MD5 of items)
    API->>API: Create cache key (safeId + contentHash)
    API->>FileSystem: Check if cached zip exists
    
    alt Cache Hit (< 1 hour old)
        FileSystem-->>API: File exists
        API-->>useDownloadCollection: {success: true, filename}
    else Cache Miss
        API->>FileSystem: Create temp directory (mkdtemp)
        API->>Archive: Initialize archiver with backpressure limit
        loop For each batch of images (size: 3)
            API->>Sanity: Fetch image URL
            API->>API: Validate URL hostname (cdn.sanity.io)
            API->>API: Extract & validate file extension
            API->>Sanity: Stream image data
            Sanity-->>API: Image stream
            API->>Archive: Append stream to archive
            API->>API: Wait for batch stream completion
        end
        Archive->>FileSystem: Write zip to temp path
        API->>FileSystem: Rename temp file to cache path (retry on lock)
        API->>FileSystem: Cleanup temp directory
        API-->>useDownloadCollection: {success: true, filename}
    end
    
    Note over useDownloadCollection,API: DOWNLOAD Phase
    useDownloadCollection->>API: POST (no mode param)
    API->>API: Validate request & auth
    API->>Sanity: Fetch collection metadata
    Sanity-->>API: Return gallery items & title
    API->>API: Generate cache key
    API->>FileSystem: Read cached zip file
    FileSystem-->>API: File stream
    API->>API: Convert Node stream to Web stream
    API-->>useDownloadCollection: ZIP file stream
    useDownloadCollection->>useDownloadCollection: Save blob as file
    useDownloadCollection->>API: POST /api/contact/audience (email)
    useDownloadCollection->>DownloadDrawer: Success
    DownloadDrawer->>User: Show success message
```

### CODE REVIEW

**Comment by Maiz27** [comment]
@greptileai

File: src/app/api/download/stream/route.ts [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+        { message: 'Collection not found or empty' },
+        { status: 404 },
+      );
+    }
+
+    // 2. Determine Cache Key (Filename)
+    // Use a hash of the content (items) + identifier to ensure invalidation when content changes
+    // This resolves the issue of serving stale zips if Sanity data updates
+    const contentHash = CryptoJS.MD5(JSON.stringify(items)).toString();
+    const safeId = CryptoJS.MD5(collectionId || slug).toString();
```
[the remark this answers is not in the document]
Comment (Maiz27) [reply src/app/api/download/stream/route.ts:162]: @greptileai This was fixed, please confirm.

File: src/app/api/download/stream/route.ts [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
```
[the remark this answers is not in the document]
Comment (Maiz27) [reply src/app/api/download/stream/route.ts:162]: @greptileai it is used in the cacheKey on line 161

File: src/app/api/download/stream/route.ts [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+        }
+
+        if (shouldRun) {
+          // FIX: Update lock file ONLY after successful cleanup to prevent blocking if cleanup fails
+          cleanupOldFiles(tempDir).then(() => {
+            try {
+              fs.writeFileSync(lockPath, '');
+            } catch (lockErr) {
+              console.warn('[Cleanup] Failed to update lock file:', lockErr);
+            }
```
[the remark this answers is not in the document]
Comment (Maiz27) [reply src/app/api/download/stream/route.ts]: @greptileai Please note this was fixed

File: src/app/api/download/stream/route.ts [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+          // FIX: Update lock file ONLY after successful cleanup to prevent blocking if cleanup fails
+          cleanupOldFiles(tempDir)
+            .then(() => {
+              try {
+                fs.writeFileSync(lockPath, '');
+              } catch (lockErr) {
+                console.warn('[Cleanup] Failed to update lock file:', lockErr);
+              }
+            })
+            .catch(console.error);
```
[the remark this answers is not in the document]
Comment (Maiz27) [reply src/app/api/download/stream/route.ts:218]: @greptileai please note this was fixed multiple times and the `.catch(console.error);` logs the error
